lambda/format-user-services/format-user-services.ts

```diff
@@ -1 +1,98 @@
-export const handler = async (): Promise<void> => {};
+import {
+  SendMessageCommand,
+  SendMessageRequest,
+  SQSClient,
+} from "@aws-sdk/client-sqs";
+import { SQSEvent, SQSRecord } from "aws-lambda";
+import type {
+  UserRecordEvent,
+  UserServices,
+  Service,
+  TxmaEventBody,
+} from "./models";
+
+const { AWS_REGION } = process.env;
+
+export const validateSQSRecord = (record: SQSRecord): SQSRecord => {
+  console.log(`Validating the SQS record body${record.body}`);
+  return record;
+};
+
+export const newServicePresenter = (txmaEventBody: TxmaEventBody): Service => ({
+  client_id: txmaEventBody.client_id,
+  count_successful_logins: 1,
+  last_accessed: txmaEventBody.timestamp,
+});
+
+export const existingServicePresenter = (
+  service: Service,
+  lastAccessed: Date
+): Service => ({
+  client_id: service.client_id,
+  count_successful_logins: service.count_successful_logins + 1,
+  last_accessed: lastAccessed,
+});
+
+export const conditionallyUpsertServiceList = (
+  matchingServices: Service[],
+  TxmaEventBody: TxmaEventBody
+): Service => ({
+  ...(!matchingServices.length
+    ? newServicePresenter(TxmaEventBody)
+    : existingServicePresenter(matchingServices[0], TxmaEventBody.timestamp)), // should only ever be one match
+});
+
+export const formatRecord = (record: UserRecordEvent) => {
+  const { TxmaEventBody } = record;
+  const { ServiceList } = record;
+  const matchingRecords = ServiceList.filter(
+    (service) => TxmaEventBody.client_id === service.client_id
+  );
+  const nonMatchingRecords = ServiceList.filter(
+    (service) => TxmaEventBody.client_id !== service.client_id
+  );
+
+  return {
+    user_id: TxmaEventBody.user.user_id,
+    services: [
+      ...[conditionallyUpsertServiceList(matchingRecords, TxmaEventBody)],
+      ...nonMatchingRecords,
+    ],
+  };
+};
+
+export const sendSqsMessage = async (
+  messageBody: UserServices[],
+  queueUrl: string
+): Promise<string | undefined> => {
+  const client = new SQSClient({ region: AWS_REGION });
+  const message: SendMessageRequest = {
+    QueueUrl: queueUrl,
+    MessageBody: JSON.stringify(messageBody),
+  };
+  const result = await client.send(new SendMessageCommand(message));
+  return result.MessageId;
+};
+
+export const handler = async (event: SQSEvent): Promise<void> => {
+  const { QUEUE_URL } = process.env;
+  if (!QUEUE_URL) {
```

Comment:
I could not for the life of me get this bit to test correctly. Could use a second pair of 👀 to see what I'm doing wrong with jest?